editor/src/core/shared/uid-utils.ts

```diff
@@ -863,3 +810,95 @@ export function getUtopiaID(element: JSXElementChild | ElementInstanceMetadata):
   }
   return element.uid
 }
+
+// The length of element UIDs generated by the editor.
+const UID_LENGTH = IS_TEST_ENVIRONMENT ? 3 : 32 // in characters
+
+/**
+ * Generate a new UID suitable for elements.
+ * The UID will try to use the `possibleUID` value if possible.
+ * The UID will be guaranteed to not clash with the other IDs in the Set passed as argument.
+ * The returned UID will be trimmed to a length of UID_LENGTH.
+ * Note: if a mock UID is available, it will have precedence and be returned immediately.
+ */
+function newElementUID(possibleUID: string, existingUIDs: Set<string>): string {
+  const mockUID = generateMockNextGeneratedUID()
+  if (mockUID != null) {
+    return mockUID
+  }
+
+  let uid = truncateUID(possibleUID)
+  // if the initial UID is empty, default it to a random one.
+  if (uid.trim().length === 0) {
+    uid = newRandomUID(possibleUID, existingUIDs)
+  }
+
+  while (existingUIDs.has(uid)) {
+    uid = newRandomUID(possibleUID, existingUIDs)
+  }
+  return uid
+}
+
+// Truncates the given UID to the length UID_LENGTH.
+function truncateUID(uid: string): string {
+  return uid.slice(0, UID_LENGTH)
+}
+
+function newRandomUID(possibleUID: string, existingUIDs: Set<string>): string {
+  return IS_TEST_ENVIRONMENT
+    ? newUIDForTests(possibleUID, existingUIDs)
+    : truncateUID(strippedUUID())
+}
+
+function newUIDForTests(possibleUID: string, existingUIDs: Set<string>): string {
+  if (possibleUID.length >= UID_LENGTH) {
+    // try to use a portion of the possible starting value
+    const maxSteps = Math.floor(possibleUID.length / UID_LENGTH)
+    for (let step = 0; step < maxSteps; step++) {
+      const substringUID = possibleUID.substring(step * UID_LENGTH, (step + 1) * UID_LENGTH)
+
+      if (!existingUIDs.has(substringUID)) {
+        return substringUID
+      }
+    }
+  }
+
+  // otherwise, build an atoz UID
+  let atozUID = 'a'.repeat(UID_LENGTH)
+  while (existingUIDs.has(atozUID)) {
+    atozUID = nextTestUID(atozUID)
+  }
+  return atozUID
+}
+
+// For test UIDs, increment the input uid deterministically ('aaab'->'aaac', 'aazz'->'abaa', etc.) while keeping the original length.
+// Non alphabetical characters will be defaulted to 'a'.
+// If the UID cannot be incremented (because it's entirely made of `z`'s), an error will be thrown.
+export function nextTestUID(uid: string): string {
+  let result = uid.split('')
```

Comment:
Possibly clearer alternative:
```suggestion
  let result = Array.from(uid)
```

Reply:
oh nice, TIL https://github.com/concrete-utopia/utopia/pull/5986/commits/653afe48c7f1d004f6bbe5e8c40aa3172c832f78